blog: New ddev share implementation (#521)
ddev/ddev.com: src/content/blog/share-providers.md

#  alt: DDEV v1.25.0 share command with cloudflared provider support
categories:
  - Announcements
  - Training
  - DevOps
---

Sharing your local development environment with clients, colleagues, or testing services has always been a valuable DDEV feature. DDEV v1.25.0 makes this easier and more flexible than ever with a complete redesign of `ddev share`. The biggest news? You can now share your projects for free using Cloudflare Tunnel—no account signup or token setup required.

## What Changed in `ddev share`

Previous versions of DDEV relied exclusively on ngrok for sharing. While ngrok remains a solid choice with advanced features, v1.25.0 introduces a modular provider system that gives you options. DDEV now ships with two built-in providers:

- **ngrok**: The traditional option (requires account and authtoken)
- **cloudflared**: A new, cost-free option using Cloudflare Tunnel (requires no account or token)

You can select providers via command flags, project configuration, or global defaults—whichever fits your workflow. Existing projects using ngrok continue working unchanged; ngrok remains the default provider.

## Free Sharing with Cloudflare Tunnel

> Cloudflare Tunnel provides production-grade infrastructure for sharing your local environments at zero cost. Getting started is as simple as:

Reviewer: ```suggestion
Cloudflare Tunnel provides production-grade infrastructure for sharing your local environments at zero cost. After [installing the client](https://developers.cloudflare.com/cloudflare-one/networks/connectors/cloudflare-tunnel/do-more-with-tunnels/local-management/create-local-tunnel/), getting started is as simple as:
```
I would add that it is necessary to install a client, and i guess by simply linking to the install page might be the easiest way to solve thi

#  alt: DDEV v1.25.0 share command with cloudflared provider support
categories:
  - Announcements
  - Training
  - DevOps
---

Sharing your local development environment with clients, colleagues, or testing services has always been a valuable DDEV feature. DDEV v1.25.0 makes this easier and more flexible than ever with a complete redesign of `ddev share`. The biggest news is that you can now share your projects for free using Cloudflare Tunnel—no account signup or token setup required.

## What Changed in `ddev share`

Previous versions of DDEV relied exclusively on ngrok for sharing. While ngrok remains a solid choice with advanced features, v1.25.0 introduces a modular provider system allowing more options and flexibility. DDEV now ships with two built-in providers:

- **ngrok**: The traditional option (requires free account and authtoken)
- **cloudflared**: A new, cost-free option using Cloudflare Tunnel (requires no account or token)

You can select providers via command flags, project configuration, or global defaults. Existing projects using ngrok continue working unchanged, and ngrok remains the default provider.

## Free Sharing with Cloudflare Tunnel

> Cloudflare Tunnel provides production-grade infrastructure for sharing your local environments at zero cost. After [installing the `cloudflared` client](https://developers.cloudflare.com/cloudflare-one/networks/connectors/cloudflare-tunnel/do-more-with-tunnels/local-management/create-local-tunnel/), getting started is:

Reviewer 2: The "download" link on this Cloudfare page goes to https://developers.cloudflare.com/cloudflare-one/networks/connectors/cloudflare-tunnel/downloads/, which might be better than the current link. Here are some advantages (+) and disadvantages (-) to changing the link:

1. (+) The current page is about creating a local tunnel, not installing `cloudflared`. That might be confusing.
2. (+) The current page has additional steps that are not needed: authenticating, creating a tunnel, configuring.
3. (-) The current page includes steps for various systems, nicely formatted. The Downloads page has some file links for Linux, but the recommend method (package manager, depending on Linux version) requires following another link to https://pkg.cloudflare.com/index.html. Even that page has more options than most users need.

If we keep the link to https://developers.cloudflare.com/cloudflare-one/networks/connectors/cloudflare-tunnel/do-more-with-tunnels/local-management/create-local-tunnel/, then one way to make it clearer would be to add the fragment `#1-download-and-install-cloudflared` and use the title of that section as the link text: [Download and install `cloudflared`](https://developers.cloudflare.com/cloudflare-one/networks/connectors/cloudflare-tunnel/do-more-with-tunnels/local-management/create-local-tunnel/#1-download-and-install-cloudflared). It should not be hard to rewrite the sentence to make that work.

```yaml
hooks:
  pre-share:
    # Make a backup of config/sites
    - exec: cp -r ${DDEV_APPROOT}/config/sites ${DDEV_APPROOT}/config/sites.bak
    - exec-host: echo "removing 'base' from site config for sharing to ${DDEV_SHARE_URL}"
    # Remove `base:` from the various site configs
    - exec: sed -i 's|^base:|#base:|g' ${DDEV_APPROOT}/config/sites/*/config.yaml
    - exec-host: echo "shared on ${DDEV_SHARE_URL}"
  post-share:
    # Restore the original configuration
    - exec: rm -rf ${DDEV_APPROOT}/config/sites
    - exec: mv ${DDEV_APPROOT}/config/sites.bak ${DDEV_APPROOT}/config/sites
    - exec-host: ddev mutagen sync
    - exec-host: echo "changes to config/sites reverted"
```

### Magento 2 Example

Magento2 has pretty easy control of the URL, so the hooks are pretty simple:

```yaml
hooks:
  pre-share:
    # Switch magento to the share URL
    - exec-host: ddev magento setup:store-config:set --base-url="${DDEV_SHARE_URL}"
  post-share:
    # Switch back to the normal local URL
    - exec-host: ddev magento setup:store-config:set --base-url="${DDEV_PRIMARY_URL}"
```

## Extensibility: Custom Share Providers

> The new provider system is script-based, allowing you to create custom providers for internal tunneling solutions or other services. Place Bash scripts in `.ddev/share-providers/` (project-level) or `~/.ddev/share-providers/` (global), and DDEV will recognize them as available providers.

Reviewer 3: ```suggestion
The new provider system is script-based, allowing you to create custom providers for internal tunneling solutions or other services. Place Bash scripts in `.ddev/share-providers/` (project-level) or `$HOME/.ddev/share-providers/` (global), and DDEV will recognize them as available providers.
```

For details on creating custom providers, see the [sharing documentation](https://docs.ddev.com/en/stable/users/topics/sharing/).

> 

Reviewer 4: ```suggestion

We love hearing what the community comes up with. Let us know on [Discord](https://discord.gg/5wjP76mBJD) what providers you created, or consider creating an addon (to share with others) or submitting a PR to include a popular provider in DDEV by default.

```

> An example of a share provider for `localtunnel` is provided in `.ddev/share-providers/localtunnel.sh.example` and you can experiment with it by just copying that to `.ddev/share-providers/localtunnel`.

Reviewer 3: ```suggestion
An example of a share provider for `localtunnel` is provided in `.ddev/share-providers/localtunnel.sh.example` and you can experiment with it by just copying that to `.ddev/share-providers/localtunnel.sh`.
```

## Questions

<dl>
<dt>Do I need to change anything in existing projects?</dt>
<dd>No. Ngrok remains the default provider, so existing projects continue working without any changes. Your ngrok authtokens and configurations are fully compatible with v1.25+.</dd>

<dt>When should I use cloudflared vs ngrok?</dt>
<dd>Use cloudflared for quick, free sharing during development and testing. Use ngrok if you need stable subdomains, custom domains, or advanced features like IP allowlisting and OAuth protection.</dd>

<dt>Can I create my own share provider?</dt>

> <dd>Yes! Place bash scripts in `.ddev/share-providers/` (project-level) or `~/.ddev/share-providers/` (global). See the <a href="https://docs.ddev.com/en/stable/users/topics/sharing/">sharing documentation</a> for implementation details.</dd>

Reviewer 3: ```suggestion
<dd>Yes! Place bash scripts in <code>.ddev/share-providers/</code> (project-level) or <code>$HOME/.ddev/share-providers/</code> (global). See the <a href="https://docs.ddev.com/en/stable/users/topics/sharing/">sharing documentation</a> for implementation details.</dd>
```

</dl>

## Try It Today

DDEV v1.25.0 is now available. Use the techniques above, and try out Cloudflared to see if you like it.

For complete details on the new sharing system, see the [sharing documentation](https://docs.ddev.com/en/stable/users/topics/sharing/).

> Join us on [Discord](https://discord.gg/5wjP76mBJD), follow us on [Mastodon](https://fosstodon.org/@ddev), [Bluesky](https://bsky.app/profile/ddev.com), or [LinkedIn](https://www.linkedin.com/company/ddev-foundation/), and subscribe to our [newsletter](https://ddev.com/newsletter/) for updates.

Reviewer 3: ```suggestion
Join us on [Discord](/s/discord), follow us on [Mastodon](https://fosstodon.org/@ddev), [Bluesky](https://bsky.app/profile/ddev.com), or [LinkedIn](https://www.linkedin.com/company/ddev-foundation/), and subscribe to our [newsletter](https://ddev.com/newsletter/) for updates.
```
